CIP-0118/README-implicit.md

```diff
@@ -0,0 +1,895 @@
+smart contracts implementing DEX functionality, or ledger changes implementing
+an on-chain order matching DEX. We argue that our solution is the most versatile,
+least expensive for users, and maintains formal guarantees
+of the current Cardano ledger, among other benefits.
+
+We discuss additional possible solutions to problems outlined
+in CPS-15 that could make use of much of the same validation zones infrastructure,
+but with varying kinds of underspecified transactions.
+
+
+## Specification
+
+
+### Validation zones
+
+A validation zone is a list of transactions. This structure replaces singleton transactions
+as a unit of validation. Specifically, instead of singleton transactions,
+validation zones are
+
+- sent across the Cardano network
+- entered into a mempool, and
+- placed in block bodies.  
+
+When validating a zone (i.e. for placement in the mempool or when validating an
+incoming block), each transaction is checked against the current ledger
+state, however, additional checks
+are done at the zone level.
+
+
+### Ledger Rule Changes
+
+The prototype for the changes can be found
+[here](https://github.com/IntersectMBO/formal-ledger-specifications/compare/polina-babel?expand=1)
+
+#### ZONE
+
+Currently, the top-level transaction-processing transition system in the ledger system is `LEDGERS`.
+It considers a list of transactions in the block body valid whenever `LEDGER` is valid for
+each transaction, checked in the order they appear in that list. We propose replacing `LEDGERS`
+with
+
+`ZONES ⊆ LEnv x LState x List (List Tx) x LState`
+
+as the top level transition, which processes each zone in the block by
+validating it with the transition
+
+`ZONE ⊆ LEnv x LState x List Tx x LState`
+
+The `ZONE` system includes two rules : `ZONE-V` may be applicable when all transactions are tagged with
+`isValid` tag that is `True`, and `ZONE-N` may be applicable otherwise. The reason
+for differentiating phase-2 valid and invalid _zones_ (and not just at the level
+of individual transactions) is that it may not be possible to apply subsequent
+transactions in a zone once a single transaction is found to be phase-2 invalid.
+Recall that  
+it is not possible to exclude only _some_ transactions from a zone in a general way,
+so, unlike the existing design, the mempool is no able to reject individual zone transactions
+dependent on a preceding transaction being phase-2 valid. We design the `ZONE-N`
+rule to deal with this situation in a special way.
+
+The following checks are done by both rules :
+
+1. the sum total of the sizes of all transactions in the zone is less than the `maxTxSize`
+protocol parameter (see [Network Requirements and Changes](#network-requirements-and-changes)), and
+2. the collateral provided by each transaction `tx` for script execution is sufficient
+to cover **all scripts in all transactions preceding `tx` in the zone, including `tx` itself**
+(see [DDoS](#ddos))
```

Comment:
This requirement severely limits the practicality of this design and IMO it is unnecessary. I think it can be treated like how I was saying to treat the [fee](https://github.com/cardano-foundation/CIPs/pull/873#discussion_r1707356525):

> The zone checks should have everything they need to calculate the total required fee for the zone, and it can compare this against the sum of the fee specified in each transaction within the zone. This would enable some transactions to have a fee of zero while still being part of a valid zone.

Since the collateral is just a multiple of the fee, the same logic applies to the collateral. Only the zone itself needs to be fully collateralized.

For the lack of practicality, imagine if two strangers each created their unbalanced transactions with the intention of them being batched with others. They don't know of each others' existence so there is no way for them to know how much collateral to include. The order of the transactions also matters, and since this is controlled by the aggregator, even if Alice and Bob new each other, they would also need to know Charlie (the aggregator) to know how much collateral to include in their own transactions.

Instead, just make collateral an all or nothing proposition. Just like how the fee is entirely taken when a zone succeeds; all collateral would be taken if a zone fails. It does not matter how much of the zone was validated before it finally failed. Doing it this way would enable strangers to use smart contracts in their own unbalanced transactions and it is up to the aggregator to provide the collateral. This is the proper alignment of incentives too since it is the aggregator that actually creates the final transaction.

Reply:
The problem with having a single collateral for a zone is, again, the fact that if someone adds a transaction with failing scripts to a zone, the collateral would be taken (again, with no detriment to the adder of the bad transaction), and whoever added that collateral will have no power to stop this. 

The collateral mechanism (for the purpose of preventing a DDoS) must guarantee that : 
- a node must be compensated for running all scripts that are run by that node as part of (phase-2) validation of an (already phase-1 valid) transaction 
- it is not possible to change the outcome of script validation for scripts in a transaction without changing and re-signing it

There is no guarantee that a single transaction is valid in a zone (since it may be unbalanced). Consider a valid zone [Tx1 ; Tx2]  : 

Tx1 is 
- unbalanced,
- executes S1, S1 succeeds
- provides collateral for only S1

Tx2 is 
- unbalanced 
- executes S2, S2 fails
- provides collateral for only S2

Zone [Tx1 ; Tx2] cannot be added as is, since Tx1 is unbalanced, and the collateral of Tx2 does not balance the zone! 

What we do instead is apply *only* the failing transaction, Tx2, when adding this zone, and only by collecting its collateral. BUT the node executed a (possibly very big and complicated S1) - who pays for this? It has to be Tx2's collateral, since no other transactions are processed in the zone. 

If we say there is a special Tx3 that provides *collateral only*, this prevents other users from adding more transactions to the zone AND this forces nodes to not validate any transactions at all in the case a zone is not complete (i.e. the collateral transaction is not yet added) - since no collateral is available for its validation.

Reply:
Note that the expected usecases here are atomic swaps, or dex aggregators, or dApp sponsorship. In all of these cases, we would expect that the user adding subsequent transactions has the capacity to provide the required collateral, and can locally check that they will not lose it no matter how zone construction proceeds thereafter.

Reply:
> The problem with having a single collateral for a zone is, again, the fact that if someone adds a transaction with failing scripts to a zone, the collateral would be taken **(again, with no detriment to the adder of the bad transaction)**, and whoever added that collateral will have no power to stop this.
-- emphasis mine

This is not true. **The only person who needs to be punished is the aggregator that batched and submitted the zone.** It is trivial for the aggregator to run the zone locally to verify the zone will succeed. There is no reason for an aggregator to ever add a transaction to the batch that will fail. IMO "the adder of the bad transaction" is **always** the aggregator. How *could* it be anyone else? The aggregator is the only entity that should put up collateral.

> Note that the expected usecases here are atomic swaps, or dex aggregators, or dApp sponsorship.

These are not the only use cases. With eUTxO style DApps, Alice may want to execute one as part of her intent. **Alice is not the aggregator** in this scenario. This is not possible with the current design because Alice does not know where in the order of the zone her intent will appear. It is literally impossible for her to know how much collateral to provide for the preceding transactions.

With the approach I mentioned, only one transaction in the zone needs to have collateral. The best transaction for this is the one the aggregator creates to finish balancing the zone. It will include enough collateral to cover **all scripts in the zone**. All transactions that come before this last transaction do not need to specify collateral. In every use case, the aggregator can add a final transaction to the zone.

> The collateral mechanism (for the purpose of preventing a DDoS) must guarantee that :
>
>  - a node must be compensated for running all scripts that are run by that node as part of (phase-2) validation of an (already phase-1 valid) transaction
> - it is not possible to change the outcome of script validation for scripts in a transaction without changing and re-signing it

Both of these guarantees are still satisfied. To use your example:
Tx1 is
- unbalanced
- executes S1, S1 succeeds
- **no collateral is specified**

Tx2 is
- unbalanced
- executes S2, S2 fails
- **provides enough collateral for S1 and S2**

In the above scenario, all of the collateral from Tx2 is taken. In an alternate scenario where S1 fails, and therefore S2 is never run, **all of the collateral from Tx2 is still taken**. It would not matter that the collateral is over-paid in the scenario that S1 fails. Again, the aggregator could have easily caught this failure by running the zone locally. Since it is their negligence, **they are the only one who should get punished**.

> BUT the node executed a (possibly very big and complicated S1) - who pays for this? It has to be Tx2's collateral, since **no other transactions are processed in the zone**. So, Tx2 has to pay for all transactions that have been validated before it.
-- emphasis mine

AFAIU the emphasized part is just an arbitrary design choice. In the case of a failed script, I see no reason why **all** transactions in the zone can't be processed in order to claim all specified collateral. No scripts need to be run in order to claim collateral. If Tx1 did specify collateral, the total collateral taken would be Tx1 collateral + Tx2 collateral, no matter which script in the zone fails.

By allowing intents to specify no collateral, **intent creators won't risk their own collateral because they are incentivized not to**. This then forces all aggregators to effectively pledge that the zone is valid since they are required to add a final transaction with the collateral even if the zone is already balanced. This is a better alignment of incentives since the aggregator is the one that actually batches and submits the zone. Whose fault is it that a node is trying to process a zone that has a failed script? This is always the aggregator's fault; they could have just omitted the intent transaction with the failing script.

Reply:
The problem I would see here is that someone could intercept the (fully valid, including scripts) zone [Tx1 ; Tx2], modify Tx1 in a way that causes S1 to fail to get Tx1', and send [Tx1' ; Tx2] across the network instead. This causes it to be entered into a block, so that, without any changes to Tx2, a malicious user has made it give up its collateral.

Reply:
Sorry, my comment was from the assumption this man-in-the-middle attack was impossible. I see know that this assumption is not accurate, as @lehins pointed out the same issue with my transaction pieces. The list of transactions itself is not being signed, but I was (ignorantly) assuming it was. AFAIU the transactions swap cip actually does sign the list of unbalanced transactions which means the above is actually possible with that design.